### PR TITLE
[test] SM Remote function tests

### PR DESCRIPTION
**Pull request by bmouryakr**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
Add sanity tests to check compatibility of sagemaker images with sagemaker remote function feature.
### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [x] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [x] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [x] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [x] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/sanity/test_remote_function_compatibility.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+@pytest.mark.model("N/A")
+@pytest.mark.integration("remote_function")
+def test_remote_function(image):
+    """
+    Test to check compatibility of sagemaker training images with sagemaker remote function.
+    """
+    python_version = get_python_version_from_image_uri(image).replace("py", "")
+    python_version = int(python_version)
+
+    if python_version < 37 or "training" not in image:
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_remote_function_compatibility.py:19]: Instead of checking for `"training" not in image`, please use `def test_remote_function(training):` instead, which ensures that only training images are passed to the function.

File: test/dlc_tests/sanity/test_remote_function_compatibility.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        pytest.skip(
+            f"Skipping remote function compatibility test for {image}. Test only for training images with Python>3.6"
+        )
+
+    repo_name, image_tag = image.split("/")[-1].split(":")
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_remote_function_compatibility.py:24]: Please use the pre-existing helper function: https://github.com/aws/deep-learning-containers/blob/03aa3331cb81696e6d4b8d048f1223bd3b249e7f/test/test_utils/__init__.py#L1842

```suggestion
    repo_name, image_tag = get_repository_and_tag_from_image_uri(image_uri)
```

File: test/dlc_tests/sanity/test_remote_function_compatibility.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    container_name = f"{repo_name}-{image_tag}-remote-function-test"
+
+    run(
+        f"docker run -itd --name {container_name} "
+        f"--mount type=bind,src=$(pwd)/container_tests,target=/test"
+        f" --entrypoint='/bin/bash' "
+        f" --env SAGEMAKER_INTERNAL_IMAGE_URI={image}"
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_remote_function_compatibility.py:31]: Why is this necessary?

File: test/dlc_tests/sanity/test_remote_function_compatibility.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        f"{image}",
+        echo=True,
+        hide=True,
+    )
+    try:
+        docker_exec_cmd = f"docker exec -i {container_name}"
+        run(f"{docker_exec_cmd} python /test/bin/test_remote_function.py ", hide=True)
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_remote_function_compatibility.py:38]: Please use the pre-existing helper function: https://github.com/aws/deep-learning-containers/blob/03aa3331cb81696e6d4b8d048f1223bd3b249e7f/test/test_utils/__init__.py#L1962

```suggestion
        run_cmd_on_container(container_name, context, "python /test/bin/test_remote_function.py")
```

File: test/dlc_tests/sanity/test_remote_function_compatibility.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import pytest
+
+from invoke import run
+from test.test_utils import get_python_version_from_image_uri
+
+
+@pytest.mark.usefixtures("sagemaker_only")
+@pytest.mark.model("N/A")
+@pytest.mark.integration("remote_function")
+def test_remote_function(image):
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_remote_function_compatibility.py:12]:
```suggestion
def test_remote_function(training):
```

File: test/dlc_tests/sanity/test_remote_function_compatibility.py (hunk 3 above)
Comment (bmouryakr) [reply test/dlc_tests/sanity/test_remote_function_compatibility.py:31]: Sagemaker SDK uses this env variable to get the default image uri for @remote decorator.

File: test/dlc_tests/sanity/test_remote_function_compatibility.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    repo_name, image_tag = image.split("/")[-1].split(":")
+    container_name = f"{repo_name}-{image_tag}-remote-function-test"
+
+    run(
+        f"docker run -itd --name {container_name} "
+        f"--mount type=bind,src=$(pwd)/container_tests,target=/test"
+        f" --entrypoint='/bin/bash' "
+        f" --env SAGEMAKER_INTERNAL_IMAGE_URI={image}"
+        f"{image}",
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_remote_function_compatibility.py:32]: Please make the spaces consistent, by adding a space at the end of each line, or at the beginning of each line, so that formatting errors are easier to avoid.

In this case, there is a space missing between `f" --env SAGEMAKER_INTERNAL_IMAGE_URI={image}"` and `f"{image}"`, which leads to errors in running the test.

File: test/dlc_tests/sanity/test_remote_function_compatibility.py (hunk 1 above)
Comment (bmouryakr) [reply test/dlc_tests/sanity/test_remote_function_compatibility.py]: Updated

File: test/dlc_tests/sanity/test_remote_function_compatibility.py (hunk 2 above)
Comment (bmouryakr) [reply test/dlc_tests/sanity/test_remote_function_compatibility.py]: Using `get_container_name` helper function instead, which covers this.

File: test/dlc_tests/sanity/test_remote_function_compatibility.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    python_version = get_python_version_from_image_uri(image).replace("py", "")
+    python_version = int(python_version)
+
+    if python_version < 37 or "training" not in image:
+        pytest.skip(
+            f"Skipping remote function compatibility test for {image}. Test only for training images with Python>3.6"
+        )
+
+    repo_name, image_tag = image.split("/")[-1].split(":")
+    container_name = f"{repo_name}-{image_tag}-remote-function-test"
```
[the remark this answers is not in the document]
Comment (bmouryakr) [reply test/dlc_tests/sanity/test_remote_function_compatibility.py]: Updated

File: test/dlc_tests/sanity/test_remote_function_compatibility.py (hunk 4 above)
Comment (bmouryakr) [reply test/dlc_tests/sanity/test_remote_function_compatibility.py]: Updated to use helper function

File: test/dlc_tests/sanity/test_remote_function_compatibility.py (hunk 5 above)
Comment (bmouryakr) [reply test/dlc_tests/sanity/test_remote_function_compatibility.py]: Updated

File: test/dlc_tests/sanity/test_remote_function_compatibility.py (hunk 6 above)
Comment (bmouryakr) [reply test/dlc_tests/sanity/test_remote_function_compatibility.py]: Thanks for catching this, corrected.

File: test/dlc_tests/container_tests/bin/test_remote_function.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    role="SageMakerRole",
+    instance_type="ml.m5.xlarge",
+    sagemaker_session=sagemaker_session,
+)
+def dlc_remote_function_test_divide(x, y):
+    return x / y
+
+
+def main():
+    dlc_remote_function_test_divide(8, 4)
```
Comment (saimidu) [inline test/dlc_tests/container_tests/bin/test_remote_function.py:25]: Do you need to check the return value for the remote function?

Comment (bmouryakr) [reply test/dlc_tests/container_tests/bin/test_remote_function.py:25]: Added assertion

**Review by saimidu — APPROVED** [review]
Looks great! Thank you for adding this test for SM Remote Function! Please revert the toml file, and we can go ahead and merge this PR.

**Comment by bmouryakr** [comment]
> Looks great! Thank you for adding this test for SM Remote Function! Please revert the toml file, and we can go ahead and merge this PR.

Done. Thanks!

**Review by saimidu — APPROVED** [review]
